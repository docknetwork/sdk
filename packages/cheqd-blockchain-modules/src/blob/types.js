import { CheqdDid, Blob } from '@docknetwork/credential-sdk/types';
import { TypedTuple } from '@docknetwork/credential-sdk/types/generic';

export class OwnerWithBlob extends TypedTuple {
  static Classes = [CheqdDid, Blob];
}
