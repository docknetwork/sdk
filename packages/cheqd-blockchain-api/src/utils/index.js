export * from './transfer';
export * from './tx';
