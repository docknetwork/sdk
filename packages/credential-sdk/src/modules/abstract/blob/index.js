export { default as AbstractBlobModule } from './module';
export * from './errors';
