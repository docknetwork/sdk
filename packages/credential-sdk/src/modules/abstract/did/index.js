export { default as AbstractDIDModule } from './module';
export * from './errors';
