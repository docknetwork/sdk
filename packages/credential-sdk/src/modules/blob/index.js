export { default as AbstractBlobModule } from './module';
