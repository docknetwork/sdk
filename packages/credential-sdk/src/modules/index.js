export * from './abstract';
export * from './multi-api';
