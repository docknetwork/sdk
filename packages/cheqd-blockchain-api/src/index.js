import { AbstractApiProvider } from '@docknetwork/credential-sdk/modules/abstract/common';
import {
  maybeToJSON,
  maybeToJSONString,
  fmtIter,
} from '@docknetwork/credential-sdk/utils';
import {
  DIDModule,
  ResourceModule,
  createCheqdSDK,
  CheqdNetwork,
} from '@cheqd/sdk';
import {
  MsgCreateDidDocPayload,
  MsgUpdateDidDocPayload,
  MsgDeactivateDidDocPayload,
  protobufPackage as didProtobufPackage,
} from '@cheqd/ts-proto/cheqd/did/v2/index.js';
import {
  MsgCreateResourcePayload,
  protobufPackage as resourceProtobufPackage,
} from '@cheqd/ts-proto/cheqd/resource/v2/index.js';
import { DidRef, NamespaceDid } from '@docknetwork/credential-sdk/types';
import { TypedEnum } from '@docknetwork/credential-sdk/types/generic';

export class CheqdAPI extends AbstractApiProvider {
  /**
   * Creates a new instance of the CheqdAPI object, call init to initialize
   * @constructor
   */
  constructor() {
    super();
  }

  static Fees = {
    MsgCreateDidDoc: DIDModule.fees.DefaultCreateDidDocFee,
    MsgUpdateDidDoc: DIDModule.fees.DefaultUpdateDidDocFee,
    MsgDeactivateDidDoc: DIDModule.fees.DefaultDeactivateDidDocFee,
    MsgCreateResource: ResourceModule.fees.DefaultCreateResourceDefaultFee,
  };

  static Prefixes = {
    MsgCreateDidDoc: didProtobufPackage,
    MsgUpdateDidDoc: didProtobufPackage,
    MsgDeactivateDidDoc: didProtobufPackage,
    MsgCreateResource: resourceProtobufPackage,
  };

  static Payloads = {
    MsgCreateDidDoc: MsgCreateDidDocPayload,
    MsgUpdateDidDoc: MsgUpdateDidDocPayload,
    MsgDeactivateDidDoc: MsgDeactivateDidDocPayload,
    MsgCreateResource: MsgCreateResourcePayload,
  };

  /**
   * Initializes `CheqdAPI` with the supplied url, wallet and network type.
   * @param {object} configuration
   * @param {string} [configuration.url]
   * @param {*} [configuration.wallet]
   * @param {string} [configuration.network]
   * @returns {this}
   */
  async init({ url, wallet, network } = {}) {
    if (network !== CheqdNetwork.Mainnet && network !== CheqdNetwork.Testnet) {
      throw new Error(
        `Invalid network provided: \`${network}\`, expected one of \`${fmtIter(
          Object.values(CheqdNetwork),
        )}\``,
      );
    } else if (wallet == null) {
      throw new Error('`wallet` must be provided');
    }

    this.ensureNotInitialized();
    const options = {
      modules: [DIDModule, ResourceModule],
      rpcUrl: url,
      wallet,
      network,
    };
    const sdk = await createCheqdSDK(options);

    this.ensureNotInitialized();
    this.sdk = sdk;

    return this;
  }

  /**
   * @returns {Promise<this>}
   */
  async disconnect() {
    this.ensureInitialized();

    delete this.sdk;

    return this;
  }

  /**
   * @returns {boolean}
   */
  isInitialized() {
    return this.sdk != null;
  }

  /**
   * Converts payload of the supplied method to bytes.
   *
   * @param {string} method
   * @param {object} payload
   * @returns {Promise<Uint8Array>}
   */
  async stateChangeBytes(method, payload) {
    const { [method]: Payload } = this.constructor.Payloads;
    if (Payload == null) {
      throw new Error(
        `Can't find payload constructor for the provided method \`${method}\``,
      );
    }
    const jsonPayload = maybeToJSON(payload);
    const sdkPayload = Payload.fromPartial(jsonPayload);

    try {
      return Payload.encode(sdkPayload).finish();
    } catch (err) {
      throw new Error(
        `Failed to encode payload \`${maybeToJSONString(sdkPayload)}\`: ${err}`,
      );
    }
  }

  /**
   * Signs and sends supplied transaction with the supplied configuration.
   *
   * @param {object} tx
   * @param {object} configuration
   * @param {string} configuration.from
   * @param {object} configuration.fee
   * @returns {Promise<*>}
   */
  async signAndSend(tx, { from, fee, memo } = {}) {
    const sender = from ?? (await this.sdk.options.wallet.getAccounts())[0].address;
    const { typeUrl } = tx;

    const prefix = this.constructor.Prefixes[typeUrl];
    const amount = fee ?? this.constructor.Fees[typeUrl];
    const payment = {
      amount: [amount],
      gas: '3600000', // TODO: dynamically calculate needed amount
      payer: sender,
    };

    const txJSON = tx.toJSON();
    txJSON.typeUrl = `/${prefix}.${typeUrl}`;

    const res = await this.sdk.signer.signAndBroadcast(
      sender,
      [txJSON],
      payment,
      memo ?? '',
    );

    if (res.code) {
      console.error(res);

      throw new Error(
        JSON.stringify(res, (_, value) => (typeof value === 'bigint' ? `${value.toString()}n` : value)),
      );
    }

    return res;
  }

  methods() {
    return ['cheqd'];
  }

  supportsIdentifier(id) {
    this.ensureInitialized();
    const { network } = this.sdk.options;

    if (id instanceof NamespaceDid) {
      if (id.isCheqd) {
        if (id.asCheqd.isTestnet) {
          return network === CheqdNetwork.Testnet;
        } else if (id.asCheqd.isMainnet) {
          return network === CheqdNetwork.Mainnet;
        }
      }
    } else if (id instanceof DidRef) {
      return this.supportsIdentifier(id[0]);
    } else if (id instanceof TypedEnum) {
      return this.supportsIdentifier(id.value);
    } else if (id?.constructor?.Qualifier?.includes(`cheqd:${network}:`)) {
      return true;
    }

    return false;
  }
}

export { CheqdNetwork };
