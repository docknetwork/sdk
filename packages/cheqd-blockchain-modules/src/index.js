import { AbstractCoreModules } from '@docknetwork/credential-sdk/modules';
import CheqdAttestModule from './attest/module';
import CheqdBlobModule from './blob/module';
import CheqdDIDModule from './did/module';
// import CheqdAccumulatorModule from './accumulator/module';
// import CheqdAnchorModule from './anchor/module';
// import CheqdOffchainSignaturesModule from './offchain-signatures/module';
// import CheqdBBSModule from './offchain-signatures/bbs';
// import CheqdBBSPlusModule from './offchain-signatures/bbs-plus';
// import CheqdPSModule from './offchain-signatures/ps';
// import CheqdStatusListCredentialModule from './status-list-credential/module';
// import CheqdTrustRegistryModule from './trust-registry/module';

export class CheqdCoreModules extends AbstractCoreModules {
  // static AccumulatorModule = CheqdAccumulatorModule;

  // static AnchorModule = CheqdAnchorModule;

  static AttestModule = CheqdAttestModule;

  static BlobModule = CheqdBlobModule;

  static DIDModule = CheqdDIDModule;

  // static OffchainSignaturesModule = CheqdOffchainSignaturesModule;

  // static BBSModule = CheqdBBSModule;

  // static BBSPlusModule = CheqdBBSPlusModule;

  // static PSModule = CheqdPSModule;

  // static StatusListCredentialModule = CheqdStatusListCredentialModule;

  // static TrustRegistryModule = CheqdTrustRegistryModule;
}

export {
  CheqdAttestModule,
  CheqdDIDModule,
  CheqdBlobModule,
  // CheqdAccumulatorModule,
  // CheqdAnchorModule,
  // ,
  // CheqdOffchainSignaturesModule,
  // CheqdBBSModule,
  // CheqdBBSPlusModule,
  // CheqdPSModule,
  // CheqdStatusListCredentialModule,
  // CheqdTrustRegistryModule,
};
