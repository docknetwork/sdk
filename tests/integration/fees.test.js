import { BTreeSet } from "@polkadot/types";
import {
  Accumulator,
  AccumulatorParams,
  initializeWasm,
  BBSPlusKeypairG2,
  PositiveAccumulator,
  BBSPlusSignatureParamsG1,
  WitnessUpdatePublicInfo,
} from "@docknetwork/crypto-wasm-ts";
import { DockAPI } from "../../src/index";
import { BN, hexToU8a, stringToHex, u8aToHex } from "@polkadot/util";
import { randomAsHex } from "@polkadot/util-crypto";
import { createNewDockDID } from "../../src/utils/did";
import { createDidPair, getBalance } from "./helpers";
import {
  createRandomRegistryId,
  OneOfPolicy,
} from "../../src/utils/revocation";
import { BLOB_MAX_BYTE_SIZE } from "../../src/modules/blob";
import { ServiceEndpointType } from "../../src/modules/did/service-endpoint";
import BBSPlusModule from "../../src/modules/bbs-plus";
import AccumulatorModule from "../../src/modules/accumulator";
import {
  FullNodeEndpoint,
  TestAccountURI,
  TestKeyringOpts,
  TestAccountCouncilMemberURI,
} from "../test-constants";

describe("Fees", () => {
  const dock = new DockAPI();

  beforeAll(async () => {
    await dock.init({
      keyring: TestKeyringOpts,
      address: FullNodeEndpoint,
    });

    await initializeWasm();
  });

  beforeEach(() => {
    const account = dock.keyring.addFromUri(TestAccountURI);
    dock.setAccount(account);
  });

  afterAll(async () => {
    await dock.disconnect();
  }, 10000);

  const withPaidFeeAssertion = async (
    sendTx,
    assert,
    applyToBefore = Object.create(null)
  ) => {
    const { address } = dock.getAccount();

    const before = await getBalance(dock.api, address);
    const result = await sendTx();
    const after = await getBalance(dock.api, address);

    // Compare free balances with `10 DOCK` precision
    assert(Math.round((before.free - after.free) / 1e7) * 10);

    for (const prop of ["feeFrozen", "reserved", "miscFrozen"]) {
      const beforeProp = applyToBefore[prop]
        ? applyToBefore[prop](before[prop])
        : before[prop];

      expect([prop, "=", after[prop].toString()]).toEqual([
        prop,
        "=",
        beforeProp.toString(),
      ]);
    }

    return result;
  };

  const withPaidFeeMatchingSnapshot = (sendTx, applyToBefore) =>
    withPaidFeeAssertion(
      sendTx,
      (fee) => expect(fee).toMatchSnapshot(),
      applyToBefore
    );

  const throwingWithPaidFeeAssertion = async (
    sendTx,
    assert,
    applyToBefore
  ) => {
    let throwed = false;
    await withPaidFeeAssertion(
      async () => {
        try {
          await sendTx();
        } catch {
          throwed = true;
        }
      },
      assert,
      applyToBefore
    );
    expect(throwed).toBe(true);
  };

  test("dids", async () => {
    const [did, pair, didKey] = createDidPair(dock);

    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.new(did, [didKey], [], false, undefined, false);
    });

    // Add DID key with all verification relationships
    const [, , dk1] = createDidPair(dock);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk1], did, did, pair, 1, undefined, false);
    });

    // Add DID key with only 1 verification relationship
    const [, , dk2] = createDidPair(dock);
    dk2.verRels.setAuthentication();
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk2], did, did, pair, 1, undefined, false);
    });

    // Add DID key with only 2 verification relationships
    const [, , dk3] = createDidPair(dock);
    dk3.verRels.setAuthentication();
    dk3.verRels.setAssertion();
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk3], did, did, pair, 1, undefined, false);
    });

    // Add DID key with 3 verification relationships
    const [, , dk4] = createDidPair(dock);
    dk4.verRels.setAuthentication();
    dk4.verRels.setAssertion();
    dk4.verRels.setCapabilityInvocation();
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk4], did, did, pair, 1, undefined, false);
    });

    // Add 2 DID keys with only 1 verification relationship
    const [, , dk5] = createDidPair(dock);
    const [, , dk6] = createDidPair(dock);
    dk5.verRels.setAuthentication();
    dk6.verRels.setCapabilityInvocation();
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk5, dk6], did, did, pair, 1, undefined, false);
    });

    // Add 3 DID keys with only 1 verification relationship
    const [, , dk7] = createDidPair(dock);
    const [, , dk8] = createDidPair(dock);
    const [, , dk9] = createDidPair(dock);
    dk7.verRels.setAuthentication();
    dk8.verRels.setCapabilityInvocation();
    dk9.verRels.setAssertion();
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys(
        [dk7, dk8, dk9],
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    const newControllers = [
      createNewDockDID(),
      createNewDockDID(),
      createNewDockDID(),
    ];
    // Add 1 controller
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addControllers(
        [newControllers[0]],
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Add 2 controllers
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addControllers(
        [newControllers[1], newControllers[2]],
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    const spType = new ServiceEndpointType();
    spType.setLinkedDomains();
    const spId1 = randomAsHex(10);
    const spId2 = randomAsHex(20);
    const origins1 = [randomAsHex(100)];
    const origins2 = [randomAsHex(100), randomAsHex(100)];
    // Add 1 service endpoint with 1 origin
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addServiceEndpoint(
        spId1,
        spType,
        origins1,
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Add 1 service endpoint with 2 origins
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addServiceEndpoint(
        spId2,
        spType,
        origins2,
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Adding a new DID which doesn't control itself but controlled by one other controller
    const [did1] = createDidPair(dock);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.new(did1, [], [did], false);
    });

    // Adding a new DID which doesn't control itself but controlled by 2 other controllers
    const [did2] = createDidPair(dock);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.new(did2, [], [did, did1], false);
    });

    // Adding a new DID which doesn't control itself but has a key and controlled by one other controller
    const [did3, , dk_] = createDidPair(dock);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.new(did3, [dk_], [did], false);
    });

    // Add DID key with all verification relationships to a DID that doesn't control itself
    const [, , dk__] = createDidPair(dock);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.addKeys([dk__], did1, did, pair, 1, undefined, false);
    });

    // Removing 1 key
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.removeKeys([2], did, did, pair, 1, undefined, false);
    });

    // Removing 2 keys
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.removeKeys([3, 4], did, did, pair, 1, undefined, false);
    });

    // Removing 1 controller
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.removeControllers(
        [newControllers[0]],
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Removing 2 controllers
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.removeControllers(
        [newControllers[1], newControllers[2]],
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Removing 1 service endpoint
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.removeServiceEndpoint(
        spId1,
        did,
        did,
        pair,
        1,
        undefined,
        false
      );
    });

    // Remove DID
    await withPaidFeeMatchingSnapshot(() => {
      return dock.did.remove(did, did, pair, 1, undefined, false);
    });
  }, 40000);

  test("revocation", async () => {
    const [did, pair, dk] = createDidPair(dock);
    await dock.did.new(did, [dk], [], false);

    const registryId = createRandomRegistryId();
    // Create owners
    const owners = new Set();
    owners.add(did);

    const policy = new OneOfPolicy(owners);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.revocation.newRegistry(registryId, policy, false, false);
    });

    let revokeIds;
    for (const count of [1, 2, 3, 5, 10]) {
      revokeIds = new BTreeSet();
      for (let i = 0; i < count; i++) {
        revokeIds.add(randomAsHex(32));
      }

      const [update, sig, nonce] = await dock.revocation.createSignedRevoke(
        registryId,
        revokeIds,
        did,
        pair,
        1,
        { didModule: dock.did }
      );
      const revTx = dock.revocation.createRevokeTx(update, [[sig, nonce]]);

      await withPaidFeeMatchingSnapshot(async () => {
        await dock.signAndSend(revTx, false);
      });
    }

    const [update, sig, nonce] = await dock.revocation.createSignedRemove(
      registryId,
      did,
      pair,
      1,
      { didModule: dock.did }
    );
    const revTx = dock.revocation.createRemoveRegistryTx(update, [
      [sig, nonce],
    ]);

    await withPaidFeeMatchingSnapshot(() => {
      return dock.signAndSend(revTx, false);
    });
  }, 40000);

  test("anchors", async () => {
    const anc = randomAsHex(32);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.anchor.deploy(anc, false);
    });
  }, 30000);

  test("blobs", async () => {
    const [did, pair, dk] = createDidPair(dock);
    await dock.did.new(did, [dk], [], false);

    const blobId = randomAsHex(32);
    const blob = {
      id: blobId,
      blob: randomAsHex(BLOB_MAX_BYTE_SIZE),
    };
    await withPaidFeeMatchingSnapshot(() => {
      return dock.blob.new(blob, did, pair, 1, { didModule: dock.did }, false);
    });
  }, 30000);

  test("bbsPlus", async () => {
    const [did, pair, dk] = createDidPair(dock);
    await dock.did.new(did, [dk], [], false);

    const label = stringToHex("My BBS+ params");

    // Add params with different attribute sizes
    for (const attributeCount of [10, 11, 12, 13, 14, 15]) {
      const bytes = u8aToHex(
        BBSPlusSignatureParamsG1.generate(attributeCount, hexToU8a(label)).toBytes()
      );
      const params = BBSPlusModule.prepareAddParameters(
        bytes,
        undefined,
        label
      );
      await withPaidFeeMatchingSnapshot(async () => {
        await dock.bbsPlusModule.addParams(
          params,
          did,
          pair,
          1,
          { didModule: dock.did },
          false
        );
      });
    }

    // Add a public key
    const kp = BBSPlusKeypairG2.generate(
      BBSPlusSignatureParamsG1.generate(10, hexToU8a(label))
    );
    const pk = BBSPlusModule.prepareAddPublicKey(
      u8aToHex(kp.publicKey.bytes),
      undefined,
      [did, 1]
    );
    await withPaidFeeMatchingSnapshot(() => {
      return dock.bbsPlusModule.addPublicKey(
        pk,
        did,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    // Remove public key
    await withPaidFeeMatchingSnapshot(() => {
      return dock.bbsPlusModule.removePublicKey(
        2,
        did,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    // Remove params
    await withPaidFeeMatchingSnapshot(() => {
      return dock.bbsPlusModule.removeParams(
        1,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });
  }, 30000);

  test("accumulator", async () => {
    const [did, pair, dk] = createDidPair(dock);
    await dock.did.new(did, [dk], [], false);

    const label = stringToHex("My Accumulator params");
    const bytes = u8aToHex(Accumulator.generateParams(hexToU8a(label)).bytes);
    const params = AccumulatorModule.prepareAddParameters(
      bytes,
      undefined,
      label
    );
    await withPaidFeeMatchingSnapshot(() => {
      return dock.accumulatorModule.addParams(
        params,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    const kp = Accumulator.generateKeypair(
      new AccumulatorParams(hexToU8a(params.bytes))
    );

    const pk = AccumulatorModule.prepareAddPublicKey(
      u8aToHex(kp.publicKey.bytes),
      undefined,
      [did, 1]
    );
    await withPaidFeeMatchingSnapshot(() => {
      return dock.accumulatorModule.addPublicKey(
        pk,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    const accumulatorPos = PositiveAccumulator.initialize(
      new AccumulatorParams(hexToU8a(params.bytes)),
      kp.secretKey
    );
    const accumulatorIdPos = randomAsHex(32);
    const accumulatedPos = u8aToHex(accumulatorPos.accumulated);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.accumulatorModule.addPositiveAccumulator(
        accumulatorIdPos,
        accumulatedPos,
        [did, 1],
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    const accumulatorIdUni = randomAsHex(32);
    const accumulatedUni = u8aToHex(accumulatorPos.accumulated);
    await withPaidFeeMatchingSnapshot(() => {
      return dock.accumulatorModule.addUniversalAccumulator(
        accumulatorIdUni,
        accumulatedUni,
        [did, 1],
        10000,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });

    const start = 10;
    // The following isn't correct logically but is good enough for getting transaction pricing
    const accumulated = u8aToHex(accumulatorPos.accumulated);
    for (let i = 1; i <= 5; i++) {
      const members = [];
      for (let j = 0; j < i; j++) {
        const member = Accumulator.encodePositiveNumberAsAccumulatorMember(
          start * 10 * i + j
        );
        members.push(member);
      }
      let witUpd = u8aToHex(
        WitnessUpdatePublicInfo.new(
          hexToU8a(accumulated),
          members,
          [],
          kp.secretKey
        ).value
      );
      await withPaidFeeMatchingSnapshot(async () => {
        await dock.accumulatorModule.updateAccumulator(
          accumulatorIdPos,
          accumulated,
          {
            additions: members.map((m) => u8aToHex(m)),
            witnessUpdateInfo: witUpd,
          },
          did,
          pair,
          1,
          { didModule: dock.did },
          false
        );
      });

      witUpd = u8aToHex(
        WitnessUpdatePublicInfo.new(
          hexToU8a(accumulated),
          [],
          members,
          kp.secretKey
        ).value
      );

      await withPaidFeeMatchingSnapshot(async () => {
        await dock.accumulatorModule.updateAccumulator(
          accumulatorIdPos,
          accumulated,
          {
            removals: members.map((m) => u8aToHex(m)),
            witnessUpdateInfo: witUpd,
          },
          did,
          pair,
          1,
          { didModule: dock.did },
          false
        );
      });
    }

    await withPaidFeeMatchingSnapshot(() => {
      return dock.accumulatorModule.removeAccumulator(
        accumulatorIdPos,
        did,
        pair,
        1,
        { didModule: dock.did },
        false
      );
    });
  }, 30000);

  test("transfers", async () => {
    const BOB = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty";

    await withPaidFeeMatchingSnapshot(() =>
      dock.signAndSend(dock.api.tx.balances.transfer(BOB, 100), false)
    );

    const txs = Array(3).fill(dock.api.tx.balances.transfer(BOB, 100));

    const txBatch = dock.api.tx.utility.batch(txs);

    await withPaidFeeMatchingSnapshot(() => dock.signAndSend(txBatch, false));
  }, 30000);

  test("democracy.notePreimage", async () => {
    const runtimeVersion = await dock.api.rpc.state.getRuntimeVersion();
    const specVersion = runtimeVersion.specVersion.toNumber();

    if (process.env.DISABLE_DEMOCRACY_PREIMAGE_TESTS || specVersion < 44) {
      return;
    }

    const preimage = randomAsHex(2e6);

    // First, send a whitelist transaction. It should have a 40K DOCK deposit and a 5K DOCK fee.
    await withPaidFeeAssertion(
      () =>
        dock.signAndSend(dock.api.tx.democracy.notePreimage(preimage), false),
      (fee) => expect(fee).toBe(45020),
      {
        // 40K DOCK per deposit
        reserved: (before) => before.add(new BN(4e10)),
      }
    );

    // Second, attempt to execute a transaction that will fail (preimage already exists). The result total fee will be around 20K, with no deposit.
    await throwingWithPaidFeeAssertion(
      () =>
        dock.signAndSend(dock.api.tx.democracy.notePreimage(preimage), false),
      (fee) => expect(fee).toBe(20020)
    );
  }, 30000);

  test("democracy.notePreimageOperational", async () => {
    const runtimeVersion = await dock.api.rpc.state.getRuntimeVersion();
    const specVersion = runtimeVersion.specVersion.toNumber();

    if (process.env.DISABLE_DEMOCRACY_PREIMAGE_TESTS || specVersion < 44) {
      return;
    }

    const preimage = randomAsHex(4e6);
    const stashAccount = dock.keyring.addFromUri(
      TestAccountCouncilMemberURI,
      null,
      "sr25519"
    );
    dock.setAccount(stashAccount);

    // First, send a whitelist transaction. It should have a 80K DOCK deposit and a 10K DOCK fee.
    await withPaidFeeAssertion(
      () =>
        dock.signAndSend(
          dock.api.tx.council.execute(
            dock.api.tx.democracy.notePreimageOperational(preimage),
            preimage.length
          ),
          false
        ),
      (fee) => expect(fee).toBe(90120),
      {
        // 80K DOCK per deposit
        reserved: (before) => before.add(new BN(8e10)),
      }
    );

    // Second, attempt to execute a transaction that will fail implicitly (preimage already exists). The result total fee will be around 20K, with no deposit.
    await withPaidFeeAssertion(
      () =>
        dock.signAndSend(
          dock.api.tx.council.execute(
            dock.api.tx.democracy.notePreimageOperational(preimage),
            preimage.length
          ),
          false
        ),
      (fee) => expect(fee).toBe(40120)
    );

    // Finally, attempt to execute a transaction that will explicitly fail (bad origin). The result total fee will be around 40K, no deposit.
    await throwingWithPaidFeeAssertion(
      () =>
        dock.signAndSend(
          dock.api.tx.democracy.notePreimageOperational(preimage),
          false
        ),
      (fee) => expect(fee).toBe(40040)
    );
  }, 30000);
});
