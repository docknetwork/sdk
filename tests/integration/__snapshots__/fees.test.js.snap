// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fees accumulator 1`] = `10`;

exports[`Fees accumulator 2`] = `10`;

exports[`Fees accumulator 3`] = `10`;

exports[`Fees accumulator 4`] = `10`;

exports[`Fees accumulator 5`] = `10`;

exports[`Fees accumulator 6`] = `10`;

exports[`Fees accumulator 7`] = `10`;

exports[`Fees accumulator 8`] = `10`;

exports[`Fees accumulator 9`] = `10`;

exports[`Fees accumulator 10`] = `10`;

exports[`Fees accumulator 11`] = `10`;

exports[`Fees accumulator 12`] = `10`;

exports[`Fees accumulator 13`] = `10`;

exports[`Fees accumulator 14`] = `10`;

exports[`Fees accumulator 15`] = `10`;

exports[`Fees anchors 1`] = `0`;

exports[`Fees bbsPlus 1`] = `10`;

exports[`Fees bbsPlus 2`] = `10`;

exports[`Fees bbsPlus 3`] = `10`;

exports[`Fees bbsPlus 4`] = `10`;

exports[`Fees bbsPlus 5`] = `10`;

exports[`Fees bbsPlus 6`] = `10`;

exports[`Fees bbsPlus 7`] = `10`;

exports[`Fees bbsPlus 8`] = `10`;

exports[`Fees bbsPlus 9`] = `0`;

exports[`Fees blobs 1`] = `90`;

exports[`Fees dids 1`] = `0`;

exports[`Fees dids 2`] = `10`;

exports[`Fees dids 3`] = `10`;

exports[`Fees dids 4`] = `10`;

exports[`Fees dids 5`] = `10`;

exports[`Fees dids 6`] = `10`;

exports[`Fees dids 7`] = `10`;

exports[`Fees dids 8`] = `10`;

exports[`Fees dids 9`] = `10`;

exports[`Fees dids 10`] = `10`;

exports[`Fees dids 11`] = `10`;

exports[`Fees dids 12`] = `0`;

exports[`Fees dids 13`] = `10`;

exports[`Fees dids 14`] = `10`;

exports[`Fees dids 15`] = `10`;

exports[`Fees dids 16`] = `10`;

exports[`Fees dids 17`] = `10`;

exports[`Fees dids 18`] = `10`;

exports[`Fees dids 19`] = `10`;

exports[`Fees dids 20`] = `10`;

exports[`Fees dids 21`] = `20`;

exports[`Fees revocation 1`] = `0`;

exports[`Fees revocation 2`] = `10`;

exports[`Fees revocation 3`] = `10`;

exports[`Fees revocation 4`] = `10`;

exports[`Fees revocation 5`] = `10`;

exports[`Fees revocation 6`] = `10`;

exports[`Fees revocation 7`] = `60`;

exports[`Fees transfers 1`] = `0`;

exports[`Fees transfers 2`] = `10`;
